### PR TITLE
Improve daemon

### PR DESCRIPTION
**Pull request by gvanrossum**
- Fix the bug where stale trees were used.
- Add --log-file flag to daemon [re]start commands.
- Refactored GC tracking code.
- Make daemon testing in misc/incremental_checker.py more robust.
- Other small cleanups.

### CODE REVIEW

**Comment by JukkaL** [comment]
I came up with a test case that triggers the use of stale trees (it should fail on master). It's probably possible to simplify it further.

```
[case testReprocessModuleEvenIfInterfaceHashDoesNotChange]
import a
import d
[file a.py]
import b
x: b.c.A
x = b.c.A()
[file b.py]
import c
[file c.py]
class A:
    x = 1
[file d.py]
import a
def f() -> None: pass
[file a.py.2]
import b
x: b.c.A
[file c.py.3]
class A:
    x = 2
[file d.py.4]
import a
def f() -> None:
    from c import A
    a.x = [A(), a.x][0]
[builtins fixtures/list.pyi]
[stale]
[rechecked a]
[stale2]
[rechecked2 c]
[stale3]
[rechecked3 d]
[out1]
[out2]
[out3]
[out4]
```

Here's the failure I get on master:

```
tmp/d.py:4: error: Incompatible types in assignment (expression has type "object", variable has type "A")
```

**Comment by gvanrossum** [comment]
Thanks! This is now ready for review. I saw your other issue (#4185), I'll look into it but it's probably orthogonal to this PR.

**Comment by gvanrossum** [comment]
Thanks for coming up with that test! It doesn't even touch the binder -- it seems the crucial part is taking the join of two types that are both `A` but from different generations of trees.

File: mypy/build.py [hunk 1]
```diff
@@ -1621,7 +1619,6 @@ def mark_interface_stale(self, *, on_errors: bool = False) -> None:
         """Marks this module as having a stale public interface, and discards the cache data."""
         self.meta = None
         self.externally_same = False
-        self.has_errors = on_errors
```
[the remark this answers is not in the document]
Comment (gvanrossum) [reply mypy/build.py]: I introduced it in the first daemon PR but it turns out I don't need a separate flag, I can just check `self.meta is None`.

File: mypy/dmypy.py [hunk 2]
```diff
@@ -570,6 +548,48 @@ def receive(sock: socket.socket) -> Any:
     return data
 
 
+class GcLogger:
```
[the remark this answers is not in the document]
Comment (gvanrossum) [reply mypy/dmypy.py]: I'll move it as soon as there's a need for it outside this module.

**Comment by gvanrossum** [comment]
The latest push should fix AppVeyor (the problem was introduced when I mistakenly removed last_manager). I've also addressed other review feedback. I'll look into the perf issues separately.

**Pull request merged**